hpx/parallel/datapar/detail/iterator_helpers.hpp

```diff
@@ -0,0 +1,312 @@
+        HPX_FORCEINLINE std::size_t data_alignment(Iter it)
+        {
+            return reinterpret_cast<std::uintptr_t>(std::addressof(*it)) &
+                (Vc::Vector<typename Iter::value_type>::MemoryAlignment - 1);
+        }
+
+        ///////////////////////////////////////////////////////////////////////
+        template <typename Iter1, typename Iter2>
+        struct iterators_datapar_compatible_impl
+        {
+            typedef typename hpx::util::decay<Iter1>::type iterator1_type;
+            typedef typename hpx::util::decay<Iter2>::type iterator2_type;
+
+            typedef Vc::Vector<
+                    typename std::iterator_traits<iterator1_type>::value_type
+                > V1;
+            typedef Vc::Vector<
+                    typename std::iterator_traits<iterator2_type>::value_type
+                > V2;
+
+            typedef std::integral_constant<bool,
+                    V1::size == V2::size &&
+                        V1::MemoryAlignment == V2::MemoryAlignment
+                > type;
+        };
+
+        template <typename Iter1, typename Iter2>
+        struct iterators_datapar_compatible
+          : iterators_datapar_compatible_impl<Iter1, Iter2>::type
+        {};
+
+        ///////////////////////////////////////////////////////////////////////
+        template <typename Iter, typename Enable = void>
+        struct iterator_datapar_compatible_impl
+          : std::false_type
+        {};
+
+        template <typename Iter>
+        struct iterator_datapar_compatible_impl<Iter,
+                typename std::enable_if<
+                    hpx::traits::is_random_access_iterator<Iter>::value
+                >::type>
+          : std::is_arithmetic<typename std::iterator_traits<Iter>::value_type>
+        {};
+
+        template <typename Iter>
+        struct iterator_datapar_compatible
+          : iterator_datapar_compatible_impl<
+                typename hpx::util::decay<Iter>::type
+            >::type
+        {};
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    namespace detail
+    {
+        template <typename Iter, typename V, typename Enable = void>
+        struct store_on_exit
+        {
+            store_on_exit(Iter const& iter)
+              : value_(std::addressof(*iter), Vc::Aligned),
+                iter_(iter)
+            {
+            }
+            ~store_on_exit()
+            {
+                value_.store(std::addressof(*iter_), Vc::Aligned);
+            }
+
+            V* operator&() { return &value_; }
+            V const* operator&() const { return &value_; }
+
+            V value_;
+            Iter iter_;
+        };
+
+        template <typename Iter, typename V>
+        struct store_on_exit<Iter, V,
+            typename std::enable_if<
+                std::is_const<
+                    typename std::iterator_traits<Iter>::value_type
+                >::value
+            >::type>
```

Comment:
This overload prevents storing the _value_ back through a const iterator. It's not about preventing to store the iterator itself.